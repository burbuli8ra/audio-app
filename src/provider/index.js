import React, { createContext, useContext, useReducer } from 'react';
import activeArtistReducer from './reducers/activeArtistReducer';
import activeAudioIndexReducer from './reducers/activeAudioIndexReducer';
import activeTrackListReducer from './reducers/activeTrackListReducer';

export const AppContext = createContext();

export const initialState = {
  activeArtist: null,
  activeAudioIndex: null,
  activeTrackList: []
};

export const mainReducer = (state, action) => ({
  ...state,
  activeArtist: activeArtistReducer(state.activeArtist, action),
  activeAudioIndex: activeAudioIndexReducer(state.activeAudioIndex, action),
  activeTrackList: activeTrackListReducer(state.activeTrackList, action)
});

export const StateProvider = ({ reducer, initialState, children }) => (
  <AppContext.Provider value={useReducer(reducer, initialState)}>
    {children}
  </AppContext.Provider>
);

export const useAppContext = () => useContext(AppContext);