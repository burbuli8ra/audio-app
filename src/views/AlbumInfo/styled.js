import styled from 'styled-components';

const Heading = styled.div`
  color: #28aa90;
  font-size: 24px;
  font-weight: bold;
  text-decoration: underline;
  text-transform: uppercase;
  margin-bottom: 20px;
`;
Heading.displayName = 'Heading';

const Cover = styled.div`
  border: 2px solid #28aa90;
  border-radius: 50%;
  height: 60px;
  width: 60px;
  overflow: hidden;

  > img {
    height: 100%;
    width: 100%;
  }
`;
Cover.displayName = 'Cover';

const Title = styled.div`
  color: #28aa90;
`;
Title.displayName = 'Title';

const Artist = styled.div`
  color: #404040;
  font-size: 14px;
`;
Artist.displayName = 'Artist';

const Details = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-left: 20px;

  ${Title} {
    margin-bottom: 4px;
  }
`;
Details.displayName = 'Details';

const Track = styled.div`
  display: flex;
  text-align: left;
`;
Track.displayName = 'Track';

const List = styled.div`
  display: inline-flex;
  flex-direction: column;

  ${Track} {
    margin-bottom: 10px;

    &:last-of-type {
      margin-bottom: 0;
    }
  }
`;
List.displayName = 'List';

export default {
  Artist,
  Cover,
  Details,
  Heading,
  List,
  Title,
  Track
};