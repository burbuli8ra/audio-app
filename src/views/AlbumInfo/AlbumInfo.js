import React, { useEffect, useState } from 'react';
import { useAppContext } from '../../provider';
import { Loader } from '../../components';
import Styled from './styled';
import albumInfoMock from '../../__mocks__/album.json';

const AlbumInfo = ({ match }) => {
  const { id } = match.params;
  const [isLoading, setIsLoading] = useState(true);
  const [albumInfo, setAlbumInfo] = useState([]);

  const [, dispatch] = useAppContext();

  const handleAlbumClick = index => {
    dispatch({
      type: 'SET_ACTIVE_INDEX',
      payload: index
    });
    dispatch({
      type: 'SET_ACTIVE_LIST',
      payload: albumInfo.tracks
    });
    dispatch({
      type: 'SET_ACTIVE_ARTIST',
      payload: albumInfo.artist
    });
  };

  // @todo replace with actual API call
  useEffect(() => {
    const timer = setTimeout(() => {
      setAlbumInfo(albumInfoMock[id]);
      setIsLoading(false);
    }, 2000);

    return () => clearTimeout(timer);
  }, [id]);

  return (
    isLoading
      ? <Loader />
      : (
        <>
          <Styled.Heading>{albumInfo.title}</Styled.Heading>
          <Styled.List>
          {albumInfo.tracks.map((track, index) => (
            <Styled.Track
              key={track.title}
              onClick={() => handleAlbumClick(index)}
            >
              <Styled.Cover>
                <img
                  src={albumInfo.cover}
                  alt={`Cover of "${albumInfo.title}" by ${albumInfo.artist}`}
                />
              </Styled.Cover>
              <Styled.Details>
                <Styled.Title>{track.title}</Styled.Title>
                <Styled.Artist>by <em>{albumInfo.artist}</em></Styled.Artist>
              </Styled.Details>
            </Styled.Track>
          ))}
          </Styled.List>
        </>
      )
  );
};

export default AlbumInfo;