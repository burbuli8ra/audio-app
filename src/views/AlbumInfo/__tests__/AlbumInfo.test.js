import React from 'react';
import { act, render } from '@testing-library/react';
import { StateProvider, initialState, mainReducer } from '../../../provider';
import AlbumInfo from '../AlbumInfo';

describe('AlbumInfo view', () => {
  let wrapper;
  const props = {
    match: {
      params: {
        id: 1
      }
    }
  };

  const setup = props => (
    <StateProvider reducer={mainReducer} initialState={initialState}>
      <AlbumInfo {...props} />
    </StateProvider>
  );

  test('should match snapshot', () => {
    jest.useFakeTimers();
    wrapper = render(setup(props));

    act(() => jest.runAllTimers());

    expect(wrapper).toMatchSnapshot();
  });
});