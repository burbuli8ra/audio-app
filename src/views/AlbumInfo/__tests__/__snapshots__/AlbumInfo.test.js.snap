// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumInfo view should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-fzpans bTxBFn"
      >
        The Villain Way
      </div>
      <div
        class="sc-fzqBZW bxtAMv"
      >
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Destroy and Blame Doggo
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Rub Face on Everything
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Licks Your Face
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Scratch at Fleas
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Meow Until Belly Rubs
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
        <div
          class="sc-fzokOt fEGKlu"
        >
          <div
            class="sc-fzplWN fmRWiL"
          >
            <img
              alt="Cover of \\"The Villain Way\\" by Bob Villain"
              src="https://www.instagram.com/p/BFmPMNpS7ig/media"
            />
          </div>
          <div
            class="sc-fznZeY iKDQWd"
          >
            <div
              class="sc-fznyAO dNAXsz"
            >
              Smelly Cat
            </div>
            <div
              class="sc-fznKkj ivYtlQ"
            >
              by 
              <em>
                Bob Villain
              </em>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-fzpans bTxBFn"
    >
      The Villain Way
    </div>
    <div
      class="sc-fzqBZW bxtAMv"
    >
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Destroy and Blame Doggo
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Rub Face on Everything
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Licks Your Face
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Scratch at Fleas
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Meow Until Belly Rubs
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
      <div
        class="sc-fzokOt fEGKlu"
      >
        <div
          class="sc-fzplWN fmRWiL"
        >
          <img
            alt="Cover of \\"The Villain Way\\" by Bob Villain"
            src="https://www.instagram.com/p/BFmPMNpS7ig/media"
          />
        </div>
        <div
          class="sc-fznZeY iKDQWd"
        >
          <div
            class="sc-fznyAO dNAXsz"
          >
            Smelly Cat
          </div>
          <div
            class="sc-fznKkj ivYtlQ"
          >
            by 
            <em>
              Bob Villain
            </em>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
